improve-ephemeral-cluster-handling.md

```diff
@@ -0,0 +1,419 @@
+  with `Incomplete` status, it waits up to a minute for the cluster to
+  disappear or have its status changed to `Running`. This accounts for clusters
+  in the process of being deleted or created. If an incomplete cluster does not
+  resolve in 1 minute, the driver pod exits.
+
+* If a driver pod looks up a cluster name and finds an ephemeral cluster, it
+  compares its own deployment value to the value of the `ephemeral` label on
+  the found cluster (as returned from oshinko-cli). If the two deployments
+  don't match, the driver pod exits.
+
+* If oshinko cannot find the deployment for the driver pod, it will create a
+  long-running cluster even if an ephemeral cluster was asked for. This covers
+  the case where a driver is started using kubernetes jobs or some other mechanism.
+
+### Explicit null values in ouput from CLI `get`
+
+Since `start.sh` parses the output of oshinko-cli, it's necessary for all
+columns printed by `get` to have an explicit value (except for potentially the
+last column). The bash script cannot be expected to figure out which fields are
+present and which are not.
+
+Consequently, null values should be added for certain columns:
+
+* <missing> will be used for services and routes that might not exist
+* <shared> will be used for long-running clusters that have no associated deployment
+
+## Alternatives / Additional (in no particular order)
+
+### Make ephemeral cluster logic with with jobs
+
+Jobs don't use replicationcontrollers, so the mechanisms we use to
+track ephemeral clusters and their associated deployments will need to be
+redesigned in the case of jobs. We also will need a scheme to handle
+parallelism, number of completions, and potentially other aspects of jobs.
+
+### Consider the exit status of spark-submit
+
+There might be some consideration of the exit status of spark-submit,
+especially when we extend this solution to jobs. If a spark application
+does not complete with success, we can imagine that it may be restarted
+even if technically the pod ran to completion. In that case we may want
+to allow an ephemeral cluster to remain
+
+### Add a CLI command to associate a driver with a long-running cluster
+
+Currently we only do association on create. However, the label mechanism
+will work to associate a driver with a long-running cluster, so we may
+want to add a CLI command to handle association/disassociation of
+drivers with long-runnign clusters.
+
+### Add a CLI command to lookup all the drivers using a shared cluster
+
+Since shared (long-running) clusters may be used by multiple drivers, given
+the above we may want to add a command to lookup all the drivers using
+a shared cluster. Or maybe it just prints as part of `get`.
+
+### Allow a driver to use more than one cluster
+
+Association currently is just a simple key/value pair. Not sure it makes
+sense for a driver to be associated with more than one cluster, but maybe.
+In that case it needs to be a list.
+
```

Comment:
this would be weird, i'm not sure how a driver would switch between contexts for different spark clusters. seems like a long shot, but i can appreciate including this for thoroughness sake.